src/bitcoin/P2tr.mo

```diff
@@ -30,4 +43,77 @@ module {
             };
         };
     };
+
+    // Create script for the given P2TR key spend address.
+    public func leafScript(bip340_spender_public_key : [Nat8]) : Result.Result<Script, Text> {
+        if (bip340_spender_public_key.size() != 32) {
+            return #err("Invalid BIP-340 public key length: expected 32 but got " # Nat.toText(bip340_spender_public_key.size()));
+        };
+        #ok([
+            // #opcode(#OP_PUSHBYTES_32) is implicit and added by the
+            // #data below
+            #data(bip340_spender_public_key),
+            #opcode(#OP_CHECKSIG),
+        ]);
+    };
+
+    public func leafHash(leaf_script : Script.Script) : [Nat8] {
+        // BIP-342 tapscript
+        let TAPROOT_LEAF_TAPSCRIPT : [Nat8] = [0xc0];
+        let script_bytes = Script.toBytes(leaf_script);
+        Hash.taggedHash(Array.flatten([TAPROOT_LEAF_TAPSCRIPT, script_bytes]), "TapLeaf");
+    };
+
+    public func tweakFromKeyAndHash(internal_key : [Nat8], hash : [Nat8]) : Result.Result<Fp.Fp, Text> {
+        if (internal_key.size() != 32) {
+            return #err("Failed to compute tweak, invalid internal key length: expected 32 but got " # Nat.toText(internal_key.size()));
+        } else if (hash.size() != 32) {
+            return #err("Failed to compute tweak, invalid hash length: expected 32 but got " # Nat.toText(hash.size()));
+        };
+
+        let tagged_hash = Hash.taggedHash(Array.flatten([internal_key, hash]), "TapTweak");
+
+        let tweak = Common.readBE256(tagged_hash, 0);
+
+        if (tweak >= Curves.secp256k1.p) {
+            return #err("Failed to compute tweak, tweak is not smaller than the curve order");
+        };
+
+        #ok(Curves.secp256k1.Fp(tweak));
+    };
+
+    public func tweakPublicKey(public_key_bip340_bytes : [Nat8], tweak : Fp.Fp) : Result.Result<PublicKey, Text> {
```

Comment:
Maybe make explicit in a comment that this corresponds to a subpart of BIP341's `taproot_tweak_pubkey`.

General question: could `tweakFromKeyAndHash` and `tweakPublicKey` also be combined into a single `taproot_tweak_pubkey(pubkey, h)`? Asked differently, what is the motivation to split it?

Reply:
added a comment

I kept them separately because I wanted to have a unit test for that (it took me much more time to implement bigger code blocks because they were hard to test and thus had more bugs) and I'm not sure if it's possible to have unit tests for private functions in motoko. At least I didn't find a way to do that. Potentially we could merge those functions to be more aligned to the spec in terms of code structure, but having more fine-grained tests also has its own advantages.